src/App.tsx

```diff
@@ -1,24 +1,360 @@
+          }, 3000);
+
+          return () => {
+            clearTimeout(timer);
+          };
+        })
+        .finally(() => {
+          setIsDisabled(false);
+          setTempTodo(null);
+        });
+    }
+  };
+
+  const deleteTodo = (id: number) => {
+    setDeleteTodoId(id);
+
+    removeTodo(id)
+      .then(() => {
+        const result = todos.filter(todo => todo.id !== id);
+
+        setTodos(result);
+        setError('');
+      })
+      .catch(() => {
+        setError('Unable to delete a todo');
+
+        const timer = setTimeout(() => {
+          setError('');
+        }, 3000);
+
+        return () => {
+          clearTimeout(timer);
+        };
+      })
+      .finally(() => {
+        setDeleteTodoId(0);
+      });
+  };
+
+  const handleChangeCompleted = (id: number, title: string) => {
+    const currentItem = todos.find(item => item.id === id);
+    const todo = {
+      id,
+      userId: USER_ID,
+      title,
+      completed: false,
+    };
+
+    setLoadingTodo(id);
+
+    updateTodo(id, todo)
+      .then((data) => {
+        const completed = todos.map(item => (
+          item.id === id
+            ? { ...data, completed: !currentItem?.completed } : item
+        ));
+
+        setTodos(completed);
+      })
+      .catch(() => {
+        setError('Unable to update a todo');
+
+        const timer = setTimeout(() => {
+          setError('');
+        }, 3000);
+
+        return () => {
+          clearTimeout(timer);
+        };
+      })
+      .finally(() => {
+        setLoadingTodo(0);
+      });
+  };
+
+  const handleClearCompleted = () => {
+    todos.forEach(item => {
+      if (item.completed) {
+        setIsLoadingCompleted(true);
+        removeTodo(item.id)
+          .then(() => {
+            const filteredTodos = todos.filter(items => !items.completed);
+
+            setTodos(filteredTodos);
+          })
+          .catch(() => {
+            setError('Unable to delete a todo');
+
+            const timer = setTimeout(() => {
+              setError('');
+            }, 3000);
+
+            return () => {
+              clearTimeout(timer);
+            };
+          })
+          .finally(() => setIsLoadingCompleted(false));
+      }
+    });
+  };
+
+  const handleCompletedAll = () => {
+    todos.forEach((item) => {
+      setIsLoadingCompleted(true);
+      updateTodo(item.id, item)
+        .then(() => {
+          const completed = todos.map(todo => (
+            { ...todo, completed: true }
+          ));
+
+          setTodos(completed);
+        })
+        .catch(() => {
+          setError('Unable to update a todo');
+
+          const timer = setTimeout(() => {
+            setError('');
+          }, 3000);
+
+          return () => {
+            clearTimeout(timer);
+          };
+        })
+        .finally(() => {
+          setIsLoadingCompleted(false);
+        });
+    });
+  };
+
+  const handleUnCompletedAll = () => {
+    todos.forEach((item) => {
```

Comment:
I guess you can try to combine this function with the previous ones as they are almost the same